src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.Stream.cs

```diff
@@ -0,0 +1,347 @@
+                {
+                    int remainingBytes = Math.Min(ODataUtf8JsonWriter.chunkSize, value.Length - i);
+                    bool isFinalBlock = false ;
+
+                    // Take a chunk of bytes from the input value.
+                    ReadOnlyMemory<byte> chunk = value.Slice(i, remainingBytes);
+
+                    // If the buffer is not empty, then we copy the bytes from the buffer
+                    // to the current chunk being processed.
+                    if (this.numBytesNotWrittenFromPreviousChunk > 0)
+                    {
+                        ReadOnlyMemory<byte> bytesNotProcessedFromPreviousChunk = this.buffer.AsMemory().Slice(0,this.numBytesNotWrittenFromPreviousChunk);
+
+                        int totalLength = bytesNotProcessedFromPreviousChunk.Length + chunk.Length;
+
+                        // Rent an array to hold bytes from both previous and current chunks.
+                        byte[] combinedArray = ArrayPool<byte>.Shared.Rent(totalLength);
+
+                        // Copy bytes from bytesNotProcessedFromPreviousChunk to the combined array
+                        bytesNotProcessedFromPreviousChunk.Span.CopyTo(combinedArray);
+
+                        // Copy bytes from chunk to the combined array, starting from the end of bytesNotProcessedFromPreviousChunk
+                        chunk.Span.CopyTo(combinedArray.AsSpan(bytesNotProcessedFromPreviousChunk.Length));
+
+                        WriteChunk(combinedArray, totalLength, isFinalBlock);
+
+                        ArrayPool<byte>.Shared.Return(combinedArray);
+                    }
+                    else
+                    {
+                        WriteChunk(chunk.Span, chunk.Length, isFinalBlock);
+                    }
+
+                    // Flush the writer if the buffer threshold is reached
+                    await this.jsonWriter.FlushIfBufferThresholdReachedAsync().ConfigureAwait(false);
+                }
+            }
+
+            /// <summary>
+            /// Writes a chunk of data, encoding it using Base64, and updates the buffer with any unprocessed bytes.
+            /// </summary>
+            /// <param name="chunk">The chunk of data to be encoded and written.</param>
+            /// <param name="isFinalBlock">A boolean indicating whether this is the final block of data.</param>
+            private void WriteChunk(ReadOnlySpan<byte> chunk, int chunkLength, bool isFinalBlock)
+            {
+                // Encode the current chunk using Base64 and write it
+                this.jsonWriter.Base64EncodeAndWriteChunk(chunk.Slice(0, chunkLength), isFinalBlock, out numBytesNotWrittenFromPreviousChunk);
+
+                if (this.numBytesNotWrittenFromPreviousChunk > 0)
+                {
+                    if (this.buffer == null)
+                    {
+                        this.buffer = ArrayPool<byte>.Shared.Rent(ODataUtf8JsonWriter.chunkSize);

```

Comment:
When/How does the rented byte array get returned?

Reply:
Is `ODataUtf8JsonWriter.chunkSize` the optimal size to rent given what we're writing into the buffer in the enclosing code block?

Reply:
The rented array gets returned when the `Dispose` method gets called. About optimal size the pool does not necessary rent the exact size that is provided for renting. It can rent a size that is slightly higher than the requested size but not lower.

Reply:
@ElizabethOkerio My concern would be that we're rented too big of a byte array. E.g. if `ODataUtf8JsonWriter.chunkSize` has a value of 1024, we could be renting a byte array that is 1024 or more bytes in size to write 3 (or a little more chars) not written from previous chunk

Reply:
Maybe the question should be why not this?
```csharp
this.buffer = ArrayPool<byte>.Shared.Rent(this.numBytesNotWrittenFromPreviousChunk);
```